[WIP] Vincent playing

vincentsarago opened the pull request:
- [x] switch to pygeos
- [x] match stac-api post_search (paging, context, bbox ...)

vincentsarago committed:
sketch

vincentsarago commented on single_file_stac_api/backend.py:
```diff
@@ -1,47 +1,71 @@
-        self.items.append(item)
+        """Insert items into the database."""
+        # re-create STRtree from items
+        raise NotImplementedError
```
For uniq insert, we will need to re-created the STRTree index

vincentsarago commented on single_file_stac_api/backend.py:
```diff
@@ -52,24 +76,84 @@ def insert_collection(self, collection: Collection):
+    items: List = attr.ib()
+    num: int = attr.ib(default=0)
+    has_next: bool = attr.ib(default=False)
+    has_previous: bool = attr.ib(default=False)
```
Simple pagination 🤷

vincentsarago commented on single_file_stac_api/backend.py:
```diff
@@ -52,24 +76,84 @@ def insert_collection(self, collection: Collection):
     """application logic"""
 
-    db: Database
+    db: Database = attr.ib(factory=Database)
```
if let without default, attrs will complain about non-default parameter being placed after defaults args

kylebarron replied:
or you can set the class to keyword only

vincentsarago committed:
make the modules more generic

vincentsarago commented:
summary of last commit changes:
- moved the cli in a `/scripts` directory
- removed `settings` in backend.py (host is an input to the class now)
- added `insert_item`  in backend.Database
- updated the pagging class to work when items is empty 🤷‍♂️ 
- ⚠️  `SingleFileClient` now takes a filepath has input 
    By definition a SingleFileClient, will alway get initialized using a file, so we don't need to use the `from_file` class nor accept an external database.
- Made `Application` more agnostic to the type of client by removing the `from_file` method. (The application could work with any kind of client 🤷‍♂️ )

vincentsarago commented on single_file_stac_api/backend.py:
```diff
@@ -52,25 +79,91 @@ def insert_collection(self, collection: Collection):
+        # Insert items and collections
+        data = SingleFileStac.parse_file(self.filepath)
+        self.db.bulk_insert_collections(data.collections)
+        self.db.bulk_insert_items(data.features)
```
Note: by design, it could work with an empty database.

vincentsarago committed:
add query

vincentsarago commented on single_file_stac_api/scripts/cli.py:
```diff
@@ -0,0 +1,21 @@
+def api(filepath, host, port):
+    """start the api."""
+    client = SingleFileClient(filepath=filepath, host=f"http://{host}:{port}")
+    app = Application(client, host=host, port=port)
```
I'm still not happy about the fact that the client (database) and the application needs to have the host information set. 

@geospatial-jeff do you think it's possible to dynamically edit the `base_url` in the Item when returning the object instead of doing it on import?

vincentsarago committed:
remove host info in database

vincentsarago committed:
moar tests

vincentsarago committed:
add viewer

vincentsarago committed:
Merge pull request #2 from vincentsarago/viewer

add viewer

vincentsarago committed:
Merge pull request #1 from vincentsarago/hostAgnostic

remove host info in database
